### PR TITLE
Download image manifests via go-containerregistry

### PR DESCRIPTION
**Pull request by albertogdd**
## Description

Refactor `GetImageVersion` function from `version` package to use [go-containerregistry](https://github.com/google/go-containerregistry) library instead.
The change also enables to use the proxy and trusted CA configured in the dynakube when pulling the manifests.

## How can this be tested?

- Running e2e tests: `make test/e2e/cloudnative/proxy`
- Installing a dynakube with proxy and trusted CA

## Checklist

- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly with a single label
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/controllers/dynakube/keychain/keychain.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package keychain
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/keychain/keychain.go:1]: this change is unrelated.

Comment (albertogdd) [reply src/controllers/dynakube/keychain/keychain.go]: Thanks, I removed it.

File: src/controllers/dynakube/version/image_hash_test.go [hunk 2]
```diff
@@ -49,3 +54,36 @@ func assertVersionStatusEquals(t *testing.T, registry *fakeRegistry, imageRef re
...
+		registryMockClient := &mocks.MockClient{}
+		registryMockClient.On("GetImageVersion", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(registry.ImageVersion{}, nil)
+		imageName := "dynatrace-operator:1.0.0"
+		dockerConfig := dockerconfig.NewDockerConfig(fake.NewClient(), dynakube.DynaKube{Spec: dynakube.DynaKubeSpec{Proxy: &dynakube.DynaKubeProxy{Value: "dummy-proxy"}}})
+
+		got, err := GetImageVersion(context.TODO(), registryMockClient, imageName, dockerConfig)
+		assert.NotNil(t, got)
+		assert.Nil(t, err)
+	})
+	// t.Run("with trustedCAs", func(t *testing.T) {
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/version/image_hash_test.go:79]: Since you're moking with client - you can remove this test.

File: src/controllers/dynakube/version/updater.go [hunk 3]
```diff
@@ -115,14 +114,10 @@ func setImageIDWithDigest( //nolint:revive
 	if canonRef, ok := ref.(reference.Canonical); ok {
 		target.ImageID = canonRef.String()
 	} else if taggedRef, ok := ref.(reference.NamedTagged); ok {
-		imageVersion, err := imageVersionFunc(ctx, imageUri, dockerCfg)
+		imageVersion, err := imageVersionFunc(ctx, &registry.GoContainerRegistryClient{}, imageUri, dockerCfg)
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/version/updater.go:117]:
```suggestion
		imageVersion, err := imageVersionFunc(ctx, &registry.Client{}, imageUri, dockerCfg)
```
Since the package already has a name registry - it's okay to name it short as Client. and rename interface to not clash with client.

File: src/controllers/dynakube/version/updater_test.go [hunk 4]
```diff
@@ -344,11 +345,11 @@ func TestUpdateVersionStatus(t *testing.T) {
 		expectedDigest := "sha256:7ece13a07a20c77a31cc36906a10ebc90bd47970905ee61e8ed491b7f4c5d62f"
 		expectedID := expectedRepo + ":tag@" + expectedDigest
 		target := status.VersionStatus{}
-		boomFunc := func(_ context.Context, imagePath string, _ *dockerconfig.DockerConfig) (ImageVersion, error) {
+		boomFunc := func(_ context.Context, _ registry.Client, imagePath string, _ *dockerconfig.DockerConfig) (registry.ImageVersion, error) {
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/version/updater_test.go:348]:
```suggestion
		boomFunc := func(_ context.Context, _ registry.ImageGetter, imagePath string, _ *dockerconfig.DockerConfig) (registry.ImageVersion, error) {
```
or something like it.

File: src/registry/client.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+	"net/http"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/opencontainers/go-digest"
+)
+
+type Client interface {
```
Comment (andriisoldatenko) [inline src/registry/client.go:14]:
```suggestion
type ImageGetter interface {
```

File: src/registry/client.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	GetImageVersion(ctx context.Context, ref name.Reference, keychain authn.Keychain, transport *http.Transport) (ImageVersion, error)
+}
+
+type ImageVersion struct {
+	Version string
+	Digest  digest.Digest
+}
+
+type GoContainerRegistryClient struct{}
```
Comment (andriisoldatenko) [inline src/registry/client.go:23]:
```suggestion
type Client struct{}
```

File: src/registry/client.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+const (
+	// VersionLabel is the name of the label used on ActiveGate-provided images.
+	VersionLabel = "com.dynatrace.build-version"
+)
+
+func (r *GoContainerRegistryClient) GetImageVersion(ctx context.Context, ref name.Reference, keychain authn.Keychain, transport *http.Transport) (ImageVersion, error) {
```
Comment (andriisoldatenko) [inline src/registry/client.go:30]:
```suggestion
func (r *Client) GetImageVersion(ctx context.Context, ref name.Reference, keychain authn.Keychain, transport *http.Transport) (ImageVersion, error) {
```

File: src/controllers/dynakube/version/testdata/certs.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package testdata
+
+const Certs = `
```
Comment (andriisoldatenko) [inline src/controllers/dynakube/version/testdata/certs.go:3]: you can add comment how did you generate it: 
```
// Generated via openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 -extensions v3_ca -keyout myCA.pem -out myCA.pem
```

File: src/controllers/dynakube/version/image_hash_test.go (hunk 2 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash_test.go]: I rewrote the test using a hardcoded certificate as we discussed!

Comment (andriisoldatenko) [reply src/controllers/dynakube/version/image_hash_test.go]: make sense.

**Review by andriisoldatenko — DISMISSED** [review]

File: src/controllers/dynakube/version/image_hash.go [hunk 9]
```diff
@@ -2,80 +2,68 @@ package version
...
-	imageReference, err := alltransports.ParseImageName(transportImageName)
+	ref, err := name.ParseReference(imageName)
 	if err != nil {
-		return ImageVersion{}, errors.WithStack(err)
+		return registry.ImageVersion{}, fmt.Errorf("parsing reference %q: %w", imageName, err)
 	}
 
-	systemContext := dockerconfig.MakeSystemContext(imageReference.DockerReference(), dockerConfig)
+	if dockerConfig.Dynakube.HasProxy() {
+		proxyUrl, err := url.Parse(dockerConfig.Dynakube.Spec.Proxy.Value)
```
Comment (aorcholski) [inline src/controllers/dynakube/version/image_hash.go:35]: [Proxy.ValueFrom](https://github.com/Dynatrace/dynatrace-operator/blob/main/src/api/v1beta1/dynakube/dynakube_types.go#L43) may also be used to define proxy

Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash.go:35]: You are totally right!!! Thank you, I missed it

Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash.go]: Fixed

File: src/controllers/dynakube/version/image_hash.go [hunk 10]
```diff
@@ -2,80 +2,72 @@ package version
...
-	inspectedImage, err := sourceImage.Inspect(context.TODO())
-	if err != nil {
-		return ImageVersion{}, errors.WithStack(err)
-	} else if inspectedImage == nil {
-		return ImageVersion{}, errors.Errorf("could not inspect image: '%s'", transportImageName)
+	rootCAs := x509.NewCertPool()
+	if ok := rootCAs.AppendCertsFromPEM(trustedCAs); !ok {
+		log.Info("failed to append custom certs!")
 	}
+	transport.TLSClientConfig.RootCAs = rootCAs
```
Comment (aorcholski) [inline src/controllers/dynakube/version/image_hash.go:70]: I suppose that "if nil" [condition](https://github.com/Dynatrace/dynatrace-operator/blob/main/src/dtclient/client.go#L204) should be used here

Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash.go:74]: Fixed [here](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/ad2c7d1a66a1b40faf3835303b4ac9e5d3703718)

Comment (andriisoldatenko) [reply src/controllers/dynakube/version/image_hash.go:74]: yeah agree with @aorcholski ☝️  good point

**Review by andriisoldatenko — APPROVED** [review]

File: src/controllers/dynakube/version/updater.go [hunk 11]
```diff
@@ -164,13 +158,10 @@ func updateVersionStatusForTenantRegistry( //nolint:revive
 		"oldVersion", target.Version)
 
 	if taggedRef, ok := ref.(reference.NamedTagged); ok {
-		imageVersion, err := imageVersionFunc(ctx, imageUri, dockerCfg)
+		imageVersion, err := imageVersionFunc(ctx, &registry.Client{}, imageUri, dockerCfg)
```
Comment (0sewa0) [inline src/controllers/dynakube/version/updater.go:161]: I would have a constructor for `&registry.Client{}`, just incase.
(nicer to look at and if the constructor's signature changes then IDEs will tell you)

File: src/controllers/dynakube/version/updater.go [hunk 12]
```diff
@@ -151,7 +146,6 @@ func updateVersionStatusForTenantRegistry( //nolint:revive
 	imageUri string,
 	imageVersionFunc ImageVersionFunc,
 	dockerCfg *dockerconfig.DockerConfig,
-	dynakube *dynatracev1beta1.DynaKube,
```
Comment (0sewa0) [inline src/controllers/dynakube/version/updater.go:154]: is `DockerConfig` still necessary?
AFAIK the `keychain` does what the `DockerConfig` is used for.
- In every change you have you just use the `DockerConfig` as a holder for the `Dynakube` and `ApiReader`, so it feels a bit redundant.

check https://github.com/Dynatrace/dynatrace-operator/blob/ad2c7d1a66a1b40faf3835303b4ac9e5d3703718/src/controllers/dynakube/version/reconciler.go#L59
I don't think that is necessary as you are doing it [here](https://github.com/Dynatrace/dynatrace-operator/blob/ad2c7d1a66a1b40faf3835303b4ac9e5d3703718/src/controllers/dynakube/version/image_hash.go#L61-L62) aswell (just better)

so for this function (and others that use the `DockerConfig` in this package) I would pass in the dynakube and apiReader explicitly as they are needed.

tldr: I would try to remove the `DockerConfig` from the `version` package in general. (other packages will do the same when they move to the new library.)

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/controllers/dynakube/version/updater.go (hunk 12 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go:154]: I agree, good point! Let me try to refactor it :+1:

File: src/controllers/dynakube/version/updater.go (hunk 11 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go]: Fixed in [this commit](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/c23e53e66bb19aaa680e8fd01e05fad805afd5a1)

File: src/controllers/dynakube/version/updater.go (hunk 12 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go:154]: Fixed in [this commit](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/c23e53e66bb19aaa680e8fd01e05fad805afd5a1)

File: src/registry/client.go [hunk 13]
```diff
@@ -27,7 +27,26 @@ const (
...
+
+func NewClientBuilder() Builder {
+	return Builder{
+		client: Client{},
+	}
+}
+
+func (builder Builder) Build() Client {
+	return builder.client
+}
```
Comment (0sewa0) [inline src/registry/client.go:42]: just this is enough: (constructor != builder-pattern)
```
func NewClient() *Client {
	return &Client{}
}
```

Lets not over do it 😅

Comment (albertogdd) [reply src/registry/client.go]: True :+1: [Fixed](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/d46e3d260e9902322dde894eb54583e4d6a318b7)

File: src/controllers/dynakube/version/updater.go [hunk 14]
```diff
@@ -115,14 +114,11 @@ func setImageIDWithDigest( //nolint:revive
 	if canonRef, ok := ref.(reference.Canonical); ok {
 		target.ImageID = canonRef.String()
 	} else if taggedRef, ok := ref.(reference.NamedTagged); ok {
-		imageVersion, err := imageVersionFunc(ctx, imageUri, dockerCfg)
+		registryClient := registry.NewClient()
+		imageVersion, err := imageVersionFunc(ctx, registryClient, imageUri, dockerCfg.RegistryAuthPath, dockerCfg.Dynakube, dockerCfg.ApiReader)
```
[the remark this answers is not in the document]
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go]: [Fixed](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/cca5a50e6d13aacd2840933ee5aca3ae22f54079)

File: src/controllers/dynakube/version/updater.go [hunk 15]
```diff
@@ -164,13 +159,11 @@ func updateVersionStatusForTenantRegistry( //nolint:revive
 		"oldVersion", target.Version)
 
 	if taggedRef, ok := ref.(reference.NamedTagged); ok {
-		imageVersion, err := imageVersionFunc(ctx, imageUri, dockerCfg)
+		registryClient := registry.NewClient()
+		imageVersion, err := imageVersionFunc(ctx, registryClient, imageUri, dockerCfg.RegistryAuthPath, dockerCfg.Dynakube, dockerCfg.ApiReader)
```
[the remark this answers is not in the document]
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go]: [Fixed](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/cca5a50e6d13aacd2840933ee5aca3ae22f54079)

File: src/controllers/dynakube/version/activegate.go [hunk 16]
```diff
@@ -4,19 +4,19 @@ import (
 	"context"
 
 	"github.com/Dynatrace/dynatrace-operator/src/api/status"
-	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1/dynakube"
-	"github.com/Dynatrace/dynatrace-operator/src/dockerconfig"
+	"github.com/Dynatrace/dynatrace-operator/src/api/v1beta1/dynakube"
```
Comment (0sewa0) [inline src/controllers/dynakube/version/activegate.go:7]: I think following the pattern of `&dynatracev1beta1.DynaKube{}` would be preferred. (using the actual api version for clarity)
also you would be avoiding shadowing the package

Comment (albertogdd) [reply src/controllers/dynakube/version/activegate.go]: [Done](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/56696004da9626423e3e56b78989950e4b3db933)

**Review by andriisoldatenko — DISMISSED** [review]
Practically LGTM, tested with my new e2e tests for proxy and custom authorities.

File: src/controllers/dynakube/version/codemodules.go [hunk 17]
```diff
@@ -57,7 +57,7 @@ func (updater *codeModulesUpdater) CheckForDowngrade(latestVersion string) (bool
 	return false, nil
 }
 
-func (updater *codeModulesUpdater) UseTenantRegistry(_ context.Context, _ *dockerconfig.DockerConfig) error {
+func (updater *codeModulesUpdater) UseTenantRegistry(_ context.Context, _ string, _ *dynatracev1beta1.DynaKube, _ client.Reader) error {
 	customVersion := updater.CustomVersion()
 	if customVersion != "" {
 		updater.dynakube.Status.CodeModules = dynatracev1beta1.CodeModulesStatus{
```
Comment (0sewa0) [inline src/controllers/dynakube/version/codemodules.go:63]: why do we pass in a `dynakube`
if the `updater` already has a `dynakube` 😅 
(I think the same could be done for the `apiReader`)

File: src/controllers/dynakube/version/image_hash.go [hunk 18]
```diff
@@ -2,80 +2,92 @@ package version
...
-	imageReference, err := alltransports.ParseImageName(transportImageName)
-	if err != nil {
-		return ImageVersion{}, errors.WithStack(err)
+func GetImageVersion( //nolint:revive // argument-limit
+	ctx context.Context,
+	registryClient registry.ImageGetter,
+	imageName string,
+	registryAuthPath string,
+	dynakube *dynatracev1beta1.DynaKube,
+	apiReader client.Reader,
```
Comment (0sewa0) [inline src/controllers/dynakube/version/image_hash.go:40]: nitpick:
I would have a different order of args
`(ctx, apiReader, registryClient,  dynakube, imageName, registryAuthPath)` 
I like to have the "simple types" at the end
- So general ordering I  ⚠️ **try**⚠️  to follow
    1.  Convention (ctx first)
    2. Clients + stuff with interfaces
    3. Data structs
    4. Simple/base types
    
(don't feel the need to refactor everything, just maybe look at the new additions and try to be consistent)

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/controllers/dynakube/version/image_hash.go (hunk 18 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash.go:40]: Thank you for the explanation :+1::+1::+1:

File: src/controllers/dynakube/version/codemodules.go (hunk 17 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/codemodules.go:63]: True. apiReader must be passed still, as the updater only contains the dtClient AFAIK

Comment (0sewa0) [reply src/controllers/dynakube/version/codemodules.go]: you could add to the struct of the updater

Comment (albertogdd) [reply src/controllers/dynakube/version/codemodules.go]: :+1: [I just did it](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/9f11156912abb79ddfcf81b6048b51bfc8055fbf)

File: src/controllers/dynakube/version/image_hash.go (hunk 18 above)
Comment (albertogdd) [reply src/controllers/dynakube/version/image_hash.go]: Done [here](https://github.com/Dynatrace/dynatrace-operator/pull/1998/commits/c9043326416b8b00f85335b458127240cfeb820a)

File: src/controllers/dynakube/version/updater.go [hunk 19]
```diff
@@ -25,10 +26,10 @@ type versionStatusUpdater interface {
 	CheckForDowngrade(latestVersion string) (bool, error)
 	LatestImageInfo() (*dtclient.LatestImageInfo, error)
 
-	UseTenantRegistry(context.Context, *dockerconfig.DockerConfig) error
+	UseTenantRegistry(context.Context, string, *dynatracev1beta1.DynaKube, client.Reader) error
 }
 
-func (reconciler *Reconciler) run(ctx context.Context, updater versionStatusUpdater, dockerCfg *dockerconfig.DockerConfig) error {
+func (reconciler *Reconciler) run(ctx context.Context, updater versionStatusUpdater, registryAuthPath string, dynakube *dynatracev1beta1.DynaKube, apiReader client.Reader) error {
```
[the remark this answers is not in the document]
Comment (albertogdd) [reply src/controllers/dynakube/version/updater.go]: Done!

**Review by 0sewa0 — APPROVED** [review]
